### PR TITLE
[kubernetes] get metrics from Kube State API

### PR DESCRIPTION
**Pull request by masci**
### What does this PR do?

Retrieve metrics from Kube State API parsing Prometheus messages from a protobuf binary stream.

### CODE REVIEW

File: utils/kubernetes/kube_state_processor.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+            tags = ['host:{}'.format(self._extract_label_value("node", metric.label))]
+            if name == 'true' and val:
+                self.kube_check.service_check(service_check_name, self.kube_check.OK, tags=tags)
+            elif name == 'false' and val:
+                self.kube_check.service_check(service_check_name, self.kube_check.CRITICAL, tags=tags)
+            elif name == 'unknown' and val:
+                self.kube_check.service_check(service_check_name, self.kube_check.UNKNOWN, tags=tags)
+
+    def kube_node_status_out_of_disk(self, message, **kwargs):
+        service_check_name = 'kube_node_status_out_of_disk'
```
Comment (hkaj) [inline utils/kubernetes/kube_state_processor.py]: same here, the name should reflect its origin

File: conf.d/kubernetes.yaml.example [hunk 2]
```diff
@@ -36,3 +36,10 @@ instances:
   #
   # enabled_gauges:
   #   - filesystem.*
+
+  # Kube State API
```
Comment (hkaj) [inline conf.d/kubernetes.yaml.example]: We should also add `kube_state_url` and explain the deal with the kube state api in a detailed comment

**Comment by remh** [comment]
@masci can you add the license headers to your new files please ?

File: utils/kubernetes/__init__.py [hunk 3]
```diff
@@ -0,0 +1,2 @@
+from .kube_state_processor import KubeStateProcessor  # noqa: F401
```
Comment (masci) [inline utils/kubernetes/__init__.py]: That was the idea, along with removing non-stateful methods from `Kubeutil` into another module for clarity. I'd open a separate PR for that

File: tests/checks/mock/test_kubernetes_state.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            '_update_metrics': mock.MagicMock(),
+            '_update_kube_state_metrics': mocked,
+        }
+        config = {'instances': [{'host': 'foo', 'kube_state_url': 'https://example.com:12345'}]}
+        self.run_check(config, force_reload=True, mocks=mocks)
+        mocked.assert_called_once()
+
+    def test__update_kube_state_metrics(self):
+        f_name = os.path.join(os.path.dirname(__file__), '..', '..', 'core', 'fixtures', 'prometheus', 'protobuf.bin')
+        f = open(f_name, 'rb')
```
Comment (remh) [inline tests/checks/mock/test_kubernetes_state.py]: nitpick but use a context manager here to make sure to close the file
```python
with open(f_name, 'rb') as f:
    mocked = mock.MagicMock(return_value=f.read())
[...]
```

File: checks.d/kubernetes_state.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        kube_state_url = instance.get('kube_state_url')
+        if kube_state_url is None:
+            raise CheckException("Unable to find kube_state_url in config file.")
+
+        try:
+            payload = self._get_kube_state(kube_state_url)
+            msg = "Got a payload of size {} from Kube State API at url:{}".format(len(payload), kube_state_url)
+            self.log.debug(msg)
+        except Exception as e:
+            self.log.error("Unable to retrieve metrics from Kube State API: {}".format(e))
```
Comment (remh) [inline checks.d/kubernetes_state.py]: You might as well raise the exception here to stop the interruption

File: checks.d/kubernetes_state.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    def _get_kube_state(self, endpoint):
+        """
+        Get metrics from the Kube State API using the protobuf format.
+        """
+        headers = {
+            'accept': 'application/vnd.google.protobuf; proto=io.prometheus.client.MetricFamily; encoding=delimited',
+            'accept-encoding': 'gzip',
+        }
+        r = requests.get(endpoint, headers=headers)
```
Comment (remh) [inline checks.d/kubernetes_state.py]: Let's do a `r.raise_for_status()` after this line

**Pull request merged**